src/drivers/absurd-sql/adapter.ts

```diff
@@ -34,18 +37,99 @@ export class DatabaseAdapter implements DatabaseAdapterInterface {
     }
   }
 
+  transaction<T>(
+    f: (_tx: Tx, setResult: (res: T) => void) => void
+  ): Promise<T | void> {
+    return new Promise((resolve, reject) => {
+      this.run({ sql: 'BEGIN' })
+        .then(() => {
+          // Committing or rolling back will be done by the wrapped transaction after which it will resolve or reject the promise.
+          // We can't commit/rollback here after executing `f` (similarly to `runInTransaction`) because `f` returns before the queries have executed (since the queries are asynchronous).
+          // Hence, we would always call `COMMIT` after executing `f` so it could be that `COMMIT` is executed before the queries from the transaction are executed.
+          // It could even be that we `COMMIT` and then one of the transaction's queries fails, the failure is caught and we call `ROLLBACK`
+          // which at that point is not possible since we already committed.
+          // This can be avoided if we were able to `await` the execution of `f` but that would require `f` to return a promise.
```

Comment:
Could you elaborate on this a little? How could this be avoided if were able to await the execution of `f`, given that we kinda do by calling `resolve` in `then`?

Reply:
In the absurd-sql driver, we have to explicitly start and end transactions.
Hence, we want to do something like this:

```
let open = true
await this.db.run('BEGIN')
try {
  await f() // call user provided function which will execute queries using callbacks
                // await has no effect here because the function does not return a promise so it will actually not await anything
               // it will directly go to the finally block and commit even though the queries are not yet executed...
} catch(err) {
  open = false
  await this.db.run('ROLLBACK')
} finally {
  if (open)
    await this.db.run('COMMIT')
}
```

When we call `await f()` (i.e. the user's transaction) it doesn't await the end of the queries because they are written using a callback style (i.e. `await` won't await the execution of the queries).